README.md
Atomify [![Build Status](https://travis-ci.org/atomify/atomify.svg)](https://travis-ci.org/atomify/atomify) [![Gitter](https://badges.gitter.im/Join Chat.svg)](https://gitter.im/atomify/atomify?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge) [![Dependencies Up-to-date](https://david-dm.org/atomify/atomify.png)](https://david-dm.org/atomify/atomify)
===============

Atomic web development - Combining the power of npm, Browserify, Rework and more to build small, fully encapsulated client side modules

## Description

Atomify provides a centralized point of access to [atomify-js](http://github.com/atomify/atomify-js) and [atomify-css](http://github.com/atomify/atomify-css) both in code and on the command line. It also offers a server with live-reload and on-the-fly bundling support to make development a breeze.

## Examples

You can learn from examples in [this repository](https://github.com/atomify/atomify-examples)

## API

`atomify(opts, cb)`

Just like its constituent pieces, atomify is a function that takes an `opts` object and a `callback` function.

### opts

**opts.js** - Options to be passed to [atomify-js](https://github.com/atomify/atomify-js#opts)

**opts.css** - Options to be passed to [atomify-css](https://github.com/atomify/atomify-css#opts)

**opts.assets** - Used to configure `opts.js.assets` and `opts.css.assets` simultaneously (and identically). See links above.

**opts.server** - Options to be passed to the development server

### callback

Just like the callbacks used by `atomify-js` and `atomify-css`, but with a third parameter to denote the type of bundle being provided. `cb(err, src, type)` where type is either `'js'` or `'css'`.

### API Example

```js
// build.js
var atomify = require('atomify')

var jsConfig = './entry.js' // shorthand for {entry: './entry.js'}

var cssConfig = {
  entry: './entry.css'
  , variables: {
    background: '#f00'
  }
}

function cb (err, src, type) {
  if (type === 'js') {
    // do something with JS source bundle
  } else {
    // do something with CSS source bundle
  }
}

atomify({js: jsConfig, css: cssConfig}, cb);
```

If you don't need to access the bundled source simply provide a file path as the output property in your options and atomify will write the file for you.

### atomify.js and atomify.css

As a convenience, you can access `atomify-js` and `atomify-css` via properties on the `atomify` function.

```js
var atomify = require('atomify')

atomify.js === require('atomify-js')
atomify.css === require('atomify-css')
```

## Development server

Atomify includes a development server that provides things like on-the-fly bundling and live reload/browser sync support to make your workflow lightning fast. `atomify.server(opts)` provides basically the same API as `atomify` itself, with a few extra options (documented below) added in. The biggest difference, of course, is that instead of writing to a file or calling a callback function, `atomify.server` responds to http requests. The server can also be configured by including a `server` property in the `opts` object when calling `atomify(opts)`.

Just like with `atomify`, the options passed to `atomify.server` are expected to have a `js` and/or `css` field. When the `entry` option of either of these is requested, the server will return the results of bundling your code. If you don't want to include the actual path to your entry file in your HTML you can also provide an `alias` option field. When the alias path is requested the server will bundle using your `entry` path.

### opts.server

You can provide server-specific options in this field.

**opts.server.port** - Port to listen on. Default: 1337

**opts.server.open** - If provided, open the URL in your default browser

**opts.server.path** - The path to open. Appended to http://localhost:port by default. Default: /default, which is a generated HTML file that includes your CSS and JS bundles automatically.

**opts.server.url** - Full URL to open instead of http://localhost:port/path

**opts.server.hostname** - Use your machine's local hostname (via Node's `os.hostname()`) instead of localhost. Ideal for accessing pages from mobile devices on the same LAN. Aliased as h.

**opts.server.lr** - Enables live-reload support by injecting the live-reload script into any HTML pages served. Supports the following sub-properties.

 * sync: Use BrowserSync. Aliased as `s`. If provided as an object will be used as the [ghostMode](https://github.com/shakyShane/browser-sync/wiki/options#ghostmode) option for BrowserSync.
 * port: Port for BrowserSync server to run on. Default: 3000
 * patterns: Globbing patterns to pass to [browsersync](https://www.npmjs.org/package/browser-sync) for watching. Default: `['**.html']` relative to `process.cwd()` as well as all files in the dependency graph of your JS and CSS bundles.
 * quiet: Suppress file change notifications on the command line. Default: false
 * verbose: Log BrowserSync's debugInfo to the console. Default: false

**opts.server.sync** - Shortcut for specifying `opts.server.lr` as `{sync: true}`.. Aliased as s.

**opts.server.st** - Options to pass to [st](https://www.npmjs.org/package/st) static file server, which is what serves all non-entry/alias requests.

**opts.server.html** - Override the default HTML served at `/default`. Pass either a filepath or a function.

If you pass a function, you'll be passed one options argument with the urls to the bundled JS and CSS. You should insert those into your HTML, and return a string.

```js
{
  server: {
    html: function html (paths, done){
      // it's important to include the body tags so that the livereload snippet from browsersync can be inserted
      var html = '<body>'
        + '<link rel="stylesheet" href="' + paths.css + '">'
        + '<h1>your current url ' + paths.request + '</h1>'
        + '<script src="' + paths.js + '"></script>'
        + '</body>'

      // you can return an error if something went wrong
      done(null, html)
    }
  }
}
```

If you pass a filepath, the bundled JS and CSS will automatically be inserted at the bottom of your file. However, you can place the strings `__ATOMIFY_CSS__` and `__ATOMIFY_JS__` when you want the relevant paths inserted to override this behavior.

**opts.server.spaMode** - When set to `true`, the default HTML will always be served. This is useful for single page apps that have a router.

## package.json config

In order to support atomify turtles all the way down, you can also specify your configuration in package.json. Simply recreate an `opts` object structure in JSON with atomify as the key. (Omit output properties if not a top level package.)

```json
{
  "atomify": {
  	 "server": {
  	 	"lr": true
  	 },
    "js": {
      "entry": "index.js",
      "output": "dist/bundle.js"
    },
    "css": {
      "entry": "index.css",
      "plugins": [
        ["rework-plugin-inline", "src/assets"],
        "rework-default-unit"
      ],
      "output": "example/bundle.css"
    }
  }
}
```

For detailed information on configuring Rework plugins in package.json see the [relevant section](https://github.com/atomify/atomify-css#packagejson-config) of the atomify-css README.

## CLI

Thanks to [subarg](https://github.com/substack/subarg), nearly everything you can do in code or JSON, you can do on the command line. JS options can be specified in a `--js, -j` subarg context and CSS options can be specified in a `--css, -c` subarg context. Server options can be specified in a `--server, -s` subarg context.

If you supply the `--debug, -d` or `--output, -o` args outside the `--js` and `--css` contexts they will apply to both JS and CSS bundles. When providing an `--output` argument that applies to both, omit the file extension and it will be applied correctly for you.

You can also configure aliases by appending them after a `:` in your entry field like `-j [ entry.js:/bundle.js ]`.

Get a complete listing of options by running `atomify --help`

### CLI Examples

```bash
atomify -j [ entry.js bundle.js ]
atomify -j [ -e entry.js -e other.js -o bundle.js -d -w ]
atomify -j [ entry.js -t funkify ] -c [ entry.css ] -o bundle
atomify -j [ src/entry.js:bundle.js ] -c [ styles/entry.css:bundle.css ] --server [ --open ]
```

Any top level args (js, css, server) passed on the command line will override the corresponding configuration defined in package.json. Non-conflicting top level items will be merged with package.json configuration.

## Install

```bash
npm install atomify
```


CHANGELOG.md
# Changelog

## 7.3.1
* Upgrade to [atomify-js 4.7.3](https://github.com/atomify/atomify-js/blob/master/CHANGELOG.md#473)

## 7.3.0
* add: now notifies both in the browser and the shell when CSS files are changed. We were already doing this for JS, so now for CSS too.

## 7.2.3
* fix: remove undocumented `resourcePaths` feature from CLI so that LESS will work. #82

## 7.2.2
* internal: fixup eslint
* internal: abstract out htmlifyError. This is nice because it means external libs can print out errors in the same way atomify does. This might be useful if you want to handle an error making custom HTML on your own.

## 7.2.1
* fix: server custom html now correctly replaces JS and CSS

## 7.2.0
* add: Server can now deal with the special snowflake babel errors.
* add: Server now just logs compilation errors, instead of exiting

## 7.1.3
* fix: server: fallback if fancy error reporting fails

## 7.1.2
* fix: server: if js or css fails to compile, the error no longer throws and causes the process to exit. Instad, we're printing the error to the shell

## 7.1.1
* add: server: cache prettified error for minor speed increase
* fix: server: when showing error output in html, browser-sync now works
* internal: dependcy versions were bumped. Only minor and patch changes.

## 7.1.0
* add: server: add a "single page app mode" so all routes trigger the default HTML
* add: server: custom html function is now passed the requested url (this is handy in single page app mode)
* add: server: custom HTML gets much prettier errors, both to the shell and the browser

## 7.0.2
* fix: server: better custom default html if no option passed. #oops

## 7.0.1
* fix: server: only try to render custom default html if passed the `html` option.

## 7.0.0
* breaking: server: we no longer watch all css files under the cwd. Only the entry file and it's dependencies
* add: server: enables setting the HTML that will be served at `/default` by either file or function. Super useful if you want to server-render.
* change: sever: dramatic speed up of watching by using chokidar (built into browser-sync) instead of gaze which gets us native OS FS events were possible
* change: server: unifed logging. It's all pretty now.
* change: removed undocumented APIs around `resourcepaths`
* change: dev: now using eslint instead of jshint and jscs
* fix: server: dramatic reduction of resources by not watching node_modules
* fix: server: no longer running three file watchers, sometimes with overlapping files. We should have a much smaller memory and CPU footprint.

## 6.1.0
* Added css & assets watch mode (@serapath)

## 6.0.4
* fix race condition in server, should prevent random server crashes
* revert gaze to 0.5, because 0.6 is the suck

## 6.0.3
Fix missing color module

## 6.0.2
Update all dependencies except gaze because the latest gaze breaks on linux.

## 6.0.1
Bug fix: `hbsfy` is no longer a peer dependency because of the `atomify-js` upgrade.

## 6.0.0
Upgrade to [atomify-js 4.0.0](https://github.com/atomify/atomify-js/blob/master/CHANGELOG.md#400)



test/index.js
var test = require('tape')
  , atomify = require('../')
  , path = require('path')
  , jsf = path.join(__dirname, 'fixtures', 'js/')
  , cssf = path.join(__dirname, 'fixtures', 'css/')
  , fs = require('fs')
  , dep = require('./fixtures/js/dep.js')
  , read = function (file) {
    return fs.readFileSync(file, 'utf8')
  }

var jsCfg = {
  entry: path.join(jsf, '/entry.js')
}

var cssCfg = {
  entry: path.join(cssf, '/entry.css')
}

test('js only', function (t) {
  t.plan(3)

  atomify({js: jsCfg}, function (err, src, type) {
    t.error(err, 'does not error')
    t.ok(src.toString().indexOf(dep) > -1, 'output contains the dep')
    t.equal(type, 'js')
  })
})

test('css only', function (t) {
  t.plan(3)

  atomify({css: cssCfg}, function (err, src, type) {
    t.error(err, 'does not error')
    t.equal(src, read(path.join(cssf, 'bundle.css')))
    t.equal(type, 'css')
  })
})

test('js and css', function (t) {
  t.plan(4)

  atomify({js: jsCfg, css: cssCfg}, function (err, src, type) {
    if (src.toString().indexOf(dep) > -1) {
      t.error(err, 'js does not error')
      t.equal(type, 'js')
    }
    if (src === read(path.join(cssf, 'bundle.css'))) {
      t.error(err, 'css does not error')
      t.equal(type, 'css')
    }
  })
})

test('js output does not prevent callback', function (t) {
  t.plan(4)

  atomify({js: {
    entry: jsCfg.entry
    , output: path.join(jsf, 'bundle-output.js')
  }, css: cssCfg}, function (err, src, type) {
    if (type === 'css'){
      t.error(err, 'does not error')
      t.equal(type, 'css')
    }
    if (type === 'js'){
      t.error(err, 'does not error')
      t.equal(type, 'js')
    }
  })
})

test('css output does prevent callback', function (t) {
  t.plan(2)

  atomify({js: jsCfg, css: {
    entry: cssCfg.entry
    , output: path.join(cssf, 'bundle-output.css')
  }}, function (err, src, type) {
    if (type === 'js'){
      t.error(err, 'does not error')
      t.equal(type, 'js')
    }
  })
})

test('js property', function (t) {
  t.plan(3)

  atomify.js(jsCfg.entry, function (err, src, type) {
    t.error(err, 'does not error')
    t.ok(src.toString().indexOf(dep) > -1, 'output contains the depedency')
    t.equal(typeof type, 'undefined')
  })
})

test('css property', function (t) {
  t.plan(3)

  atomify.css(cssCfg.entry, function (err, src, type) {
    t.error(err, 'does not error')
    t.equal(src, read(path.join(cssf, 'bundle.css')))
    t.equal(typeof type, 'undefined')
  })
})

test('port conflict detection', function (t) {
  t.plan(1)

  t.throws(function () {
    atomify({
      server: {port: 3000, lr: true}
    })
  })
})


tests/entry.js
var dependant = require('dependant')
  , nondependant = require('nondependant')

alert('DONE')


tests/index.js
var atomify = require('../')
  , http = require('http')
  , path = require('path')

var cssVariables = {
  background: '#f00'
}

var atom = atomify({
  js: {
    entry: './entry.js'
  , shim: {
      jquery: { path: path.resolve('./jquery.js'), exports: '$' }
    }
  }
, css: {
    entry: './entry.css'
  , variables: cssVariables
  }
});

http.createServer(function (req, res) {

  console.log('requesting', req.url)

  if (req.url == '/style.css') atom.css(req, res)
  else if (req.url == '/index.js') atom.js(req, res)
  else {
    res.writeHead(404, {'Content-Type': 'text/plain'});
    res.end('Not Found\n');
  }

}).listen(8080);
console.log('Test server running at http://localhost:8080/');


lib/cli.js
var js = require('atomify-js')
  , css = require('atomify-css')
  , fs = require('fs')
  , path = require('path')
  , writer = require('write-to-path')
  , server = require('./server')
  , pkgCfg = path.join(process.cwd(), 'package.json')

pkgCfg = fs.existsSync(pkgCfg) && require(pkgCfg).atomify ? require(pkgCfg).atomify : {}

module.exports = function (args) {
  if ((args._ && args._[0] === 'help') || args.help) {
    return fs.createReadStream(__dirname + '/usage.txt')
      .pipe(process.stdout)
      .on('close', function () { process.exit(1) });
  }

  if (args.version) {
    return console.log(require('../package.json').version);
  }

  args.css = args.css || pkgCfg.css || fs.existsSync('./index.css') && {entry: 'index.css'}
  args.js = args.js || pkgCfg.js || fs.existsSync('./index.js') && {entry: 'index.js'}
  args.server = args.server || pkgCfg.server

  if (pkgCfg && pkgCfg.assets) {
    if (args.css) args.css.assets = pkgCfg.assets
    if (args.js) args.js.assets = pkgCfg.assets
  }

  if (args.css) {
    args.css = parseArgs('css', args.css, args)

    args.css.variables = args.css.v || args.css.variables
    args.css.plugins = args.css.p || args.css.plugins
    args.css.compress = args.css.c || args.css.compress

    if (args.css.output) {
      css(args.css, function doWatchWriteServerstart(err, src) {
        if (err) console.error(err)
        if (args.server) server(args)
      });
    }
  }

  if (args.js) {
    args.js = parseArgs('js', args.js, args)

    args.js.watch = args.js.w || args.js.watch
    args.js.transforms = args.js.t || args.js.transform

    if (args.js.output) js(args.js, writer(args.js.output, args.js))
  }
}

function parseArgs (ext, child, parent) {
  if (typeof child === 'string') child = {entry: child} // -j entry.js
  if (child._) child.entry = child._[0] // -j [ entry.js ]
  if (typeof child.e === 'string') child.entry = child.e // -j [ -e entry.js ]
  if (Array.isArray(child.e)) child.entries = child.e // -j [ -e entry.js -e other.js ]

  if (child.entry.indexOf(':') > 0) {
    var pieces = child.entry.split(':')
    child.entry = pieces[0]
    child.alias = pieces[1]
  }

  child.debug = child.d || child.debug || parent.debug

  child.output = child.o || child.output
  if (child._ && child._.length === 2) child.output = child._[1] // -j [ entry.js bundle.js ]
  if (!child.output && parent.output) child.output = parent.output.indexOf(ext) > 1 ? parent.output : parent.output + '.' + ext

  if (!child.output && !parent.server) {
    console.error('No output path provided for ' + ext.toUpperCase() + ' bundle!')
    process.exit(1)
  }

  return child;
}


lib/htmlifyError.js
'use strict'

var internals = {}
  , AnsiToHTML = require('ansi-to-html')
  , prettyError = require('prettify-error')

internals.htmlifyBabelError = function htmlifyBabelError (err, options) {
  var stack = (new AnsiToHTML({
    escapeXML: true
    , fg: '#000'
    , bg: '#fff'
  })).toHtml(err.codeFrame)
    , html = options.head
      ? '<!doctype html><html><head>'
        + '<meta charset="utf-8">'
        + '<meta http-equiv="X-UA-Compatible" content="IE=edge">'
        + '<meta name="viewport" content="initial-scale=1,width=device-width,user-scalable=0">'
        + '<title>' + err.toString() + '</title>'
        + '</head><body>'
      : ''

  html +=
    +'<div style="font-family: \'helvetica-neue\', helvetica">'
    + '<h1>' + err.toString() + '</h1>'
    + '</div>'
    + '<pre><code>' + stack + '</code></pre>'

  if (options.head) html += '</body></html>'

  return html
}

module.exports = function htmlifyError (err, options) {
  var lines
    , html
    , prettyifiedErr

  options || (options = {})

  if (err._babel) return internals.htmlifyBabelError(err, options)

  prettyifiedErr = prettyError(err) || err

  lines = (new AnsiToHTML({escapeXML: true})).toHtml(prettyifiedErr).split('\n')
  html = options.head
    ? '<!doctype html><html><head>'
      + '<meta charset="utf-8">'
      + '<meta http-equiv="X-UA-Compatible" content="IE=edge">'
      + '<meta name="viewport" content="initial-scale=1,width=device-width,user-scalable=0">'
      + '<title>' + err.toString() + '</title>'
      + '</head><body>'
    : ''

  // first line is the error
  html += '<div style="font-family: \'helvetica-neue\', helvetica">'
    + '<h1>' + lines.shift() + '</h1>'
    + '<pre><code>'
    // second line is a location marker
    + lines.shift() + '\n'
    // third through fifth lines are code
    + lines.shift() + '\n'
    + lines.shift() + '\n'
    + lines.shift() + '\n'
    // sixth line is a location marker
    + lines.shift() + '\n'
    + '</code></pre>'

    + '<ul style="list-style: none; padding: 0; margin: 0">'
      + lines.map(function eachLine (line) {
        return '<li style="margin-bottom: .5em">'
          + line.trim()
            .replace(/^at\s/, '')
            .replace(/\((.*)\)$/, '<span style="color: grey">$1</span>')
          + '</li>'
      }).join('')
    + '</ul>'
    + '</div>'

  if (options.head) html += '</body></html>'

  return html
}


lib/server.js
'use strict'

var st = require('st')
  , http = require('http')
  , path = require('path')
  , url = require('url')
  , fs = require('fs')
  , os = require('os')
  , js = require('atomify-js')
  , css = require('atomify-css')
  , prettyError = require('prettify-error')
  , ansiStyleFormat = require('style-format')
  , cssFiles = require('atomify-css/css')
  , lessFiles = require('atomify-css/less')
  , open = require('open')
  , browserSync = require('browser-sync').create()
  , htmlifyError = require('./htmlifyError.js')
  , defaultBundlePath = path.join(__dirname, '_bundle.js')
  , JS_REPLACE = '__ATOMIFY_JS__'
  , CSS_REPLACE = '__ATOMIFY_CSS__'
  , HTML_HEAD = '<!doctype html><html><head>'
    + '<meta charset="utf-8">'
    + '<meta http-equiv="X-UA-Compatible" content="IE=edge">'
    + '<meta name="viewport" content="initial-scale=1,width=device-width,user-scalable=0">'
  , cwd = process.cwd()
  , internals = {}
  , baseUrl
  , bs

internals.getDefaultHTML = function getDefaultHTML (args) {
  var src = HTML_HEAD
  src += '<title>generated by atomify</title>'
  if (args.css.entry) src += '<link rel="stylesheet" href="CSS">'
  src += '</head><body>'
  src += '<script src="JS"></script>'
  src += '</body></html>'

  src = src.replace('JS', args.js.alias)
  src = src.replace('CSS', args.css.alias)

  return src
}

internals.customDefaultHTML = ''

internals.setCustomDefaultHTML = function setCustomDefaultHTML (file, args) {
  var html = fs.readFileSync(file, {encoding: 'utf8'})
    , shouldReplaceJS = html.indexOf(JS_REPLACE) > -1
    , shouldReplaceCSS = html.indexOf(CSS_REPLACE) > -1

  if (shouldReplaceCSS) html = html.replace(CSS_REPLACE, args.css.alias)
  else html += '<link rel="stylesheet" href="' + args.css.alias + '">'

  if (shouldReplaceJS) html = html.replace(JS_REPLACE, args.js.alias)
  else html += '<script src="' + args.js.alias + '"></script>'

  internals.customDefaultHTML = html

  return internals.customDefaultHTML
}

internals.getCustomDefaultHTML = function getCustomDefaultHTML (req, args, callback) {
  var html = args.server.html
    , htmlIsFile = typeof html === 'string'

  // bail early if there is no html option
  if (!html) return void callback(null, null)

  // if the html arg is a string, assume it's a file name
  // if we've already cached this file, don't go read it again
  if (htmlIsFile && html !== internals.customDefaultHTML) {
    callback(null, internals.setCustomDefaultHTML(html, args))
  }
  else if (htmlIsFile){
    callback(null, internals.customDefaultHTML)
  }
  // if the html arg isn't a string, it should be a function that returns html
  else {
    html({js: args.js.alias, css: args.css.alias, request: req.url}, callback)
  }
}

internals.logError = function logError (err) {
  var prettyifiedErr = prettyError(err)

  // if any of our custom error parsing fails, don't die
  try {
    // babel errors are really special, so give them custom error handling
    if (err._babel) {
      bs.logger.error(ansiStyleFormat('{red}' + err.toString() + '{reset}'))
      console.error(err.codeFrame)
    }
    else {
      bs.logger.error(prettyifiedErr || err)
    }
  }
  catch (e) {
    console.error(err)
  }
}

internals.sendError = function sendError (err, res) {
  if (!res.headersSent) {
    res.setHeader('Content-Type', 'text/html')
    res.statusCode = 500
  }

  internals.logError(err)

  try {
    res.end(htmlifyError(err, {head: true}))
  }
  catch (e){
    res.end('<pre>' + err.toString() + err.stack + '</pre>')
  }

  return true
}

internals.serveDefaultPage = function serveDefaultPage (req, res, args) {
  internals.getCustomDefaultHTML(req, args, function gotCustomDefaultHTML (err, html) {
    var src = html || internals.getDefaultHTML(args)

    res.setHeader('Content-Type', 'text/html')

    if (err) internals.sendError(err, res)
    else res.end(src)
  })
}

internals.parseSyncOptions = function parseSyncOptions (opts) {
  if (!opts) return false
  if (typeof opts === 'object') return opts

  return {
    clicks: true
    , location: true
    , forms: true
    , scroll: true
  }
}

internals.getNormalizedPath = function getNormalizedPath (pathStr) {
  if (!pathStr || pathStr.charAt(0) === '/') return pathStr

  return '/' + pathStr
}

internals.responder = function responder (type, res) {
  return function response (err, src) {
    if (err) return void internals.sendError(err, res)

    if (!res.headersSent) res.setHeader('Content-Type', 'text/' + type)
    res.end(src || err)
  }
}

internals.startFileWatch = function startFileWatch (lr, args) {
  var watchedFiles = []
    , addFile = function addFile (file) {
      if (watchedFiles.indexOf(file) < 0){
        watchedFiles.push(file)
        browserSync.watch(file).on('change', function onCSSChanged () {
          bs.logger.info('{cyan:CSS changed: {magenta:%s', path.relative(cwd, file))
          browserSync.notify('CSS changed')
          browserSync.reload(args.css.alias)
        })
      }
    }

  if (!lr.quiet) {
    js.emitter.on('changed', function onJsChanged (filepath) {
      bs.logger.info('{cyan:JS changed: {magenta:%s', path.relative(cwd, filepath))
      browserSync.notify('Compiling, please wait!')
    })
  }

  js.emitter.on('bundle', function onJSBundled (time) {
    if (!lr.quiet) {
      bs.logger.info('{grey:Bundle updated in {magenta:%s', time + 'ms')
    }
    // for JS, just reload the whole page
    browserSync.reload()
  })

  cssFiles.emitter.on('file', addFile)
  lessFiles.emitter.on('file', addFile)
  addFile(args.css.entry)
}

module.exports = function server (args) {
  // sometimes the hostname has a '.local' on it, sometimes it doesn't
  var hostname = args.server.hostname || args.server.h ? os.hostname().replace(/\.local$/, '') + '.local' : 'localhost'

  args.server.url = args.server.url || 'http://' + hostname + ':1337/default'

  var stOptions = args.server.st || {path: cwd, cache: false}
    , mount = st(stOptions)
    , parsedUrl = url.parse(args.server.url)
    , port = args.server.port || parsedUrl.port
    , serverPath = internals.getNormalizedPath(args.server.path || parsedUrl.path)
    , lr = args.server.lr || args.server.l
    , launch = args.server.open || args.server.o
    , handler = args.server.handler || Function()
    , spaMode = args.server.spaMode

  if (args.server.sync || args.server.s) lr = {sync: true}

  args.servePath = stOptions.path

  // the router checks these, so they have to exist
  args.js = args.js || {}
  args.js.alias = internals.getNormalizedPath(args.js.alias || args.js.entry)
  args.css = args.css || {}
  args.css.alias = internals.getNormalizedPath(args.css.alias || args.css.entry)

  baseUrl = parsedUrl.protocol + '//' + parsedUrl.hostname

  // cache the browserify and watchify instances
  js.emitter.on('browserify', function storeBrowserifyInstace (b) {
    internals.browserify = b
  })
  js.emitter.on('watchify', function storeWatchifyInstance (w) {
    internals.watchify = w
  })

  if (lr) {
    // if live reloading, use watchify
    args.js.watch = true
    args.js.output = args.js.output || defaultBundlePath

    lr = typeof lr === 'boolean' ? {} : lr
    lr.patterns = lr.patterns
      ? lr.patterns
      : [path.join(args.servePath, '**/*.html')]

    if (args.css.output) lr.patterns.push(args.css.output)

    lr.port = lr.port || 3000
    lr.sync = internals.parseSyncOptions(lr.sync || lr.s)

    if (port >= lr.port && port <= lr.port + 2) {
      throw new Error('Ports ' + lr.port + ' through ' + (lr.port + 2) + 'in use for Live Reload')
    }

    bs = browserSync.init({
      proxy: {
        target: parsedUrl.hostname + ':' + port
      }
      , port: lr.port
      , injectChanges: true
      , ghostMode: lr.sync
      , notify: !!lr.quiet
      , logSnippet: false
      , open: false
      , logLevel: lr.verbose ? 'debug' : 'info'
      , logFileChanges: !!lr.quiet
      , logConnections: !!lr.quiet
      // don't minify for speed!
      , minify: false
      , startPath: '/default'
      , logPrefix: 'atomify'
      , files: lr.patterns.map(function doNotWatchNodeModules (match) {
        return {
          match: match
          , options: {
            ignored: /node_modules/
          }
        }
      })
    })

    internals.startFileWatch(lr, args)

    // listen for initial bundle completion before opening
    js(args.js, function onJSError (err) {
      if (err) internals.logError(err)
    }).on('end', function onJSBundled () {
      if (launch && bs.options && bs.options.port){
        open(baseUrl + ':' + bs.options.port + serverPath)
      }
    })
  }

  http.createServer(function createServer (req, res) {
    switch (req.url.split('?')[0]) {
      case args.js.alias || args.js.entry:
        if (lr) {
          res.setHeader('Content-Type', 'text/javascript')
          res.end(fs.readFileSync(args.js.output, {encoding: 'utf8'}))
        }
        else {
          js(args.js, internals.responder('javascript', res))
        }
      break

      case args.css.alias || args.css.entry:
        css(args.css, internals.responder('css', res))
      break

      case '/default':
        internals.serveDefaultPage(req, res, args)
      break

      default:
        if (spaMode) internals.serveDefaultPage(req, res, args)
        else if (handler(req, res, args)) return null
        else mount(req, res)
      break
    }
  }).listen(port)

  // if live reload is enabled we open after initial bundling
  if (launch && !lr) open(baseUrl + ':' + port + serverPath)
}

